### PR TITLE
[CELEBORN-796] Support for globally disable thread-local cache in the shared PooledByteBufAllocator

### PR DESCRIPTION
**Pull request by cfmcgrady**
### What changes were proposed in this pull request?

As title

### Why are the changes needed?

As title

### Does this PR introduce _any_ user-facing change?

Yes, the thread local cache of shared `PooledByteBufAllocator` can be disabled by setting `celeborn.network.memory.allocator.allowCache=false`

### How was this patch tested?

Pass GA

### CODE REVIEW

**Comment by pan3793** [comment]
Will disable cache hurt performance and increase cpu usage?

**Comment by pan3793** [comment]
Let's revise this PR, it's beneficial to release the direct memory.

I hit worker memory issues recently, the worker triggers trim action, then the disk buffer become empty but direct memory does not reach the threshold, causing the worker being a PAUSE state forever.

```
MemoryManager#183 - Direct memory usage: 1536.0 MiB/3.0 GiB, disk buffer size: 0.0 B, sort memory size: 0.0 B, read buffer size: 0.0 B
StorageManager#51 - Trigger org.apache.celeborn.service.deploy.worker.storage.StorageManager trim action
StorageManager#51 - Trigger org.apache.celeborn.service.deploy.worker.storage.StorageManager trim action
...
```

File: common/src/main/java/org/apache/celeborn/common/network/util/NettyUtils.java [hunk 1]
```diff
@@ -42,7 +42,8 @@
 
 /** Utilities for creating various Netty constructs based on whether we're using EPOLL or NIO. */
 public class NettyUtils {
-  private static volatile PooledByteBufAllocator _allocator;
+  private static final PooledByteBufAllocator[] _sharedPooledByteBufAllocator =
```
Comment (pan3793) [inline common/src/main/java/org/apache/celeborn/common/network/util/NettyUtils.java:45]: how about splitting it into 2 variables?

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -1232,6 +1235,15 @@ object CelebornConf extends Logging {
       .timeConf(TimeUnit.MILLISECONDS)
       .createWithDefaultString("10s")
 
+  val NETWORK_MEMORY_ALLOCATOR_ALLOW_CACHE: ConfigEntry[Boolean] =
+    buildConf("celeborn.network.memory.allocator.allowCache")
+      .categories("network")
+      .internal
+      .version("0.4.0")
+      .doc("when false, globally disabled thread-local cache in the shared PooledByteBufAllocator.")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1243]:
```suggestion
      .doc("When false, globally disable thread-local cache in the shared PooledByteBufAllocator.")
```

**Comment by pan3793** [comment]
cc @waitinfuture

**Comment by pan3793** [comment]
I think it's good to go as it does not change the default behavior.

Generally disabling cache would hurt performance, any chance to provide a benchmark with cache enabled/disabled? e.g. terasort https://github.com/pan3793/spark-terasort

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
BTW, I rolled this patch on a small cluster, it shows helps with memory reduction without 
obvious performance dropping.

Before (cache enabled)
image

After (cache disabled)
image

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 3]
```diff
@@ -1232,6 +1235,15 @@ object CelebornConf extends Logging {
       .timeConf(TimeUnit.MILLISECONDS)
       .createWithDefaultString("10s")
 
+  val NETWORK_MEMORY_ALLOCATOR_ALLOW_CACHE: ConfigEntry[Boolean] =
+    buildConf("celeborn.network.memory.allocator.allowCache")
+      .categories("network")
+      .internal
+      .version("0.4.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1242]: maybe 0.3.1, if @waitinfuture not oppose to

File: common/src/main/java/org/apache/celeborn/common/network/util/NettyUtils.java (hunk 1 above)
Comment (pan3793) [reply common/src/main/java/org/apache/celeborn/common/network/util/NettyUtils.java:45]: not a big deal, both are fine.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 3 above)
Comment (waitinfuture) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1242]: No opposition:)

Comment (waitinfuture) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1242]: The graph looks great!

**Comment by cfmcgrady** [comment]
the netty memory grows more slowly than before in our production environment after applying this patch.

Before this PR (cache enabled)

截屏2023-08-09 上午11 04 17

After this PR (cache disabled)

截屏2023-08-09 上午11 04 54

**Comment by waitinfuture** [comment]
Thanks! Merging to main/0.3

**Pull request closed without merging**